catalog/resource_sql_table.go

```diff
@@ -0,0 +1,383 @@
+}
+
+func (ti *SqlTableInfo) serializeColumnInfo(col SqlColumnInfo) string {
+	notNull := ""
+	if !col.Nullable {
+		notNull = " NOT NULL"
+	}
+
+	comment := ""
+	if col.Comment != "" {
+		comment = fmt.Sprintf(" COMMENT %s", col.Comment)
+	}
+	return fmt.Sprintf("%s %s%s%s", col.Name, col.TypeText, notNull, comment) // id INT NOT NULL COMMENT something
+}
+
+func (ti *SqlTableInfo) serializeColumnInfos() string {
+	columnFragments := make([]string, len(ti.ColumnInfos))
+	for i, col := range ti.ColumnInfos {
+		columnFragments[i] = ti.serializeColumnInfo(col)
+	}
+	return strings.Join(columnFragments[:], ", ") // id INT NOT NULL, name STRING, age INT
+}
+
+func (ti *SqlTableInfo) serializeProperties() string {
+	propsMap := make([]string, 0, len(ti.Properties))
+	for key, value := range ti.Properties {
+		if !sqlTableIsManagedProperty(key) {
+			propsMap = append(propsMap, fmt.Sprintf("'%s'='%s'", key, value))
+		}
+	}
+	return strings.Join(propsMap[:], ", ") // 'foo'='bar', 'this'='that'
+}
+
+func (ti *SqlTableInfo) buildLocationStatement() string {
+	statements := make([]string, 0, 10)
+	statements = append(statements, fmt.Sprintf("LOCATION '%s'", ti.StorageLocation)) // LOCATION '/mnt/csv_files'
+
+	if ti.StorageCredentialName != "" {
+		statements = append(statements, fmt.Sprintf(" WITH (CREDENTIAL `%s`)", ti.StorageCredentialName))
+	}
+	return strings.Join(statements, "")
+}
+
+func (ti *SqlTableInfo) getTableTypeString() string {
+	if ti.TableType == "VIEW" {
+		return "VIEW"
+	}
+	return "TABLE"
+}
+
+func (ti *SqlTableInfo) buildTableCreateStatement() string {
+	statements := make([]string, 0, 10)
+
+	isView := ti.TableType == "VIEW"
+
+	externalFragment := ""
+	if ti.TableType == "EXTERNAL" {
+		externalFragment = "EXTERNAL "
+	}
+
+	createType := ti.getTableTypeString()
+
+	statements = append(statements, fmt.Sprintf("CREATE %s%s %s", externalFragment, createType, ti.FullName()))
+
+	if len(ti.ColumnInfos) > 0 {
+		statements = append(statements, fmt.Sprintf(" (%s)", ti.serializeColumnInfos()))
+	}
+
+	if !isView {
+		if ti.DataSourceFormat != "" {
+			statements = append(statements, fmt.Sprintf("\nUSING %s", ti.DataSourceFormat)) // USING CSV
+		}
+	}
+
+	if ti.Comment != "" {
+		statements = append(statements, fmt.Sprintf("\nCOMMENT '%s'", ti.Comment)) // COMMENT 'this is a comment'
+	}
+
+	if len(ti.Properties) > 0 {
+		statements = append(statements, fmt.Sprintf("\nTBLPROPERTIES (%s)", ti.serializeProperties())) // TBLPROPERTIES ('foo'='bar')
+	}
+
+	if !isView {
+		if ti.StorageLocation != "" {
+			statements = append(statements, "\n"+ti.buildLocationStatement())
+		}
+	} else {
+		statements = append(statements, fmt.Sprintf("\nAS %s", ti.ViewDefinition))
+	}
+
+	statements = append(statements, ";")
+
+	return strings.Join(statements, "")
+}
+
+func (ti *SqlTableInfo) updateTable(ctx context.Context, d *schema.ResourceData, s map[string]*schema.Schema, c *common.DatabricksClient, oldti *SqlTableInfo) error {
```

Comment:
`d`, `s`, and `c` are not used in the body of this method. also, it's easier to support in the long term, if you rewrite this method to something like: `func (ti *SqlTableInfo) Diff(old *SqlTableInfo) ([]string, err)`, which would return a list of SQL ALTER statements, which would subsequently be applied.

Reply:
agreed